File: variance_analysis.py
#!/usr/bin/env python3
"""
Variance Analysis for Multi-Round Biclique Counting
Analyzes which (u,w) vertex pairs contribute most to variance
"""

import sys
import os
import numpy as np
from collections import defaultdict
import matplotlib.pyplot as plt

def parse_edge_list(filename):
    """Parse bipartite graph from edge list"""
    edges = []
    upper_vertices = set()
    lower_vertices = set()
    
    try:
        with open(filename, 'r') as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith('#'):
                    continue
                parts = line.split()
                if len(parts) >= 2:
                    u, v = int(parts[0]), int(parts[1])
                    edges.append((u, v))
                    upper_vertices.add(u)
                    lower_vertices.add(v)
    except Exception as e:
        print(f"Error reading file: {e}")
        return None, None, None
    
    return edges, upper_vertices, lower_vertices

def analyze_pair_variance(edges, upper_vertices, lower_vertices, num_rounds=10):
    """
    Analyze variance contribution of each (u,w) pair
    For each pair, simulate multiple rounds and track variance
    """
    
    print("\n=== Variance Analysis for (u,w) Pairs ===")
    print(f"Upper vertices: {len(upper_vertices)}")
    print(f"Lower vertices: {len(lower_vertices)}")
    print(f"Edges: {len(edges)}")
    print(f"Running {num_rounds} rounds\n")
    
    # Build adjacency lists
    upper_neighbors = defaultdict(set)
    lower_neighbors = defaultdict(set)
    
    for u, v in edges:
        upper_neighbors[u].add(v)
        lower_neighbors[v].add(u)
    
    # For each (u,w) pair, estimate butterflies across rounds
    pair_estimates = defaultdict(list)
    
    for round_num in range(num_rounds):
        print(f"Round {round_num + 1}/{num_rounds}...", end=' ', flush=True)
        
        # For each upper vertex u
        for u in upper_vertices:
            neighbors_u = upper_neighbors[u]
            
            # For each lower vertex w
            for w in lower_vertices:
                neighbors_w = lower_neighbors[w]
                
                # Count common neighbors (butterflies involving this pair)
                # A butterfly (2,2)-biclique is: u-v1-w-v2-u
                # Common neighbors of u and w in opposite partition
                common = len(neighbors_u & neighbors_w)
                
                # Estimate with noise (simulate DP)
                noise = np.random.laplace(0, 1.0)
                estimate = max(0, common + noise)
                
                pair_estimates[(u, w)].append(estimate)
        
        print("done")
    
    # Compute variance for each pair
    pair_variances = []
    
    for (u, w), estimates in pair_estimates.items():
        if len(estimates) < 2:
            continue
        
        mean = np.mean(estimates)
        variance = np.var(estimates, ddof=1)
        std_dev = np.std(estimates, ddof=1)
        
        pair_variances.append({
            'u': u,
            'w': w,
            'variance': variance,
            'mean': mean,
            'std_dev': std_dev,
            'estimates': estimates
        })
    
    # Sort by variance (descending)
    pair_variances.sort(key=lambda x: x['variance'], reverse=True)
    
    # Print top contributors
    print("\n=== Top 20 Vertex Pairs by Variance ===")
    print(f"{'Rank':<6} {'u':<8} {'w':<8} {'Variance':<15} {'Std Dev':<15} {'Mean Est':<15}")
    print("-" * 70)
    
    for i, pv in enumerate(pair_variances[:20]):
        print(f"{i+1:<6} {pv['u']:<8} {pv['w']:<8} {pv['variance']:<15.6f} {pv['std_dev']:<15.6f} {pv['mean']:<15.6f}")
    
    # Statistics
    print("\n=== Variance Statistics ===")
    total_variance = sum(pv['variance'] for pv in pair_variances)
    max_variance = max(pv['variance'] for pv in pair_variances)
    min_variance = min(pv['variance'] for pv in pair_variances)
    avg_variance = total_variance / len(pair_variances)
    
    print(f"Total pairs analyzed: {len(pair_variances)}")
    print(f"Average variance: {avg_variance:.6f}")
    print(f"Max variance: {max_variance:.6f}")
    print(f"Min variance: {min_variance:.6f}")
    
    # Top 10% contribution
    top_10_percent = max(1, len(pair_variances) // 10)
    top_10_variance = sum(pv['variance'] for pv in pair_variances[:top_10_percent])
    
    print(f"\nTop 10% pairs ({top_10_percent} pairs) contribute: {top_10_variance/total_variance*100:.2f}% of total variance")
    
    # Save detailed results
    with open('pair_variance_analysis.txt', 'w') as f:
        f.write("Vertex Pair Variance Analysis\n")
        f.write(f"Upper vertices: {len(upper_vertices)}, Lower vertices: {len(lower_vertices)}\n")
        f.write(f"Edges: {len(edges)}\n")
        f.write(f"Rounds: {num_rounds}\n\n")
        f.write("u\tw\tVariance\tStdDev\tMeanEst\n")
        
        for pv in pair_variances:
            f.write(f"{pv['u']}\t{pv['w']}\t{pv['variance']:.6f}\t{pv['std_dev']:.6f}\t{pv['mean']:.6f}\n")
    
    print("\nDetailed results saved to: pair_variance_analysis.txt")
    
    return pair_variances

def suggest_optimizations(pair_variances, upper_vertices, lower_vertices):
    """Suggest optimizations based on variance analysis"""
    
    print("\n=== Optimization Suggestions ===")
    
    # Identify high-variance pairs
    high_var_pairs = pair_variances[:max(1, len(pair_variances) // 10)]
    high_var_u = set(pv['u'] for pv in high_var_pairs)
    high_var_w = set(pv['w'] for pv in high_var_pairs)
    
    print(f"\n1. High-variance upper vertices ({len(high_var_u)} vertices):")
    for u in sorted(high_var_u)[:10]:
        print(f"   u={u}")
    
    print(f"\n2. High-variance lower vertices ({len(high_var_w)} vertices):")
    for w in sorted(high_var_w)[:10]:
        print(f"   w={w}")
    
    print("\n3. Recommended optimizations:")
    print("   a) Allocate more privacy budget to high-variance pairs")
    print("   b) Use stratified sampling: separate high/low variance pairs")
    print("   c) Apply variance reduction (control variates, importance sampling)")
    print("   d) Adaptive epsilon allocation based on pair variance")
    print("   e) Focus refinement rounds on high-variance pairs")

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 variance_analysis.py <graph_file> [num_rounds]")
        print("  graph_file: path to bipartite graph edge list")
        print("  num_rounds: number of rounds for variance analysis (default: 10)")
        sys.exit(1)
    
    graph_file = sys.argv[1]
    num_rounds = int(sys.argv[2]) if len(sys.argv) > 2 else 10
    
    if not os.path.exists(graph_file):
        print(f"Error: File not found: {graph_file}")
        sys.exit(1)
    
    print(f"Loading graph: {graph_file}")
    edges, upper_vertices, lower_vertices = parse_edge_list(graph_file)
    
    if edges is None:
        print("Error: Failed to parse graph")
        sys.exit(1)
    
    # Run variance analysis
    pair_variances = analyze_pair_variance(edges, upper_vertices, lower_vertices, num_rounds)
    
    # Suggest optimizations
    suggest_optimizations(pair_variances, upper_vertices, lower_vertices)
    
    print("\n=== Analysis Complete ===")

if __name__ == "__main__":
    main()

File: variance_analysis.cpp
#include "bigraph.h"
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <fstream>
#include <random>
#include <ctime>

using namespace std;

// Global variables needed by bigraph
long double Eps = 1.0, Eps0 = 0.5, Eps1 = 0.25, Eps2 = 0.25;
long double p = 0.5;
int K___ = 2, P___ = 2;
long double alpha0 = 0.5, alpha1 = 0.25, alpha2 = 0.25;
vector<int> priv_deg;
int priv_dmax_1 = 0, priv_dmax_2 = 0, iteration = 0, num_rounds = 1;
long double m3__ = 0, m2__ = 0, m1__ = 0, m0__ = 0;
long double RR_time = 0, server_side_time = 0, naive_server_side = 0;
long double local_count_time = 0, deg_esti_time = 0;
long double real = 0, real_ld = 0;
long double avg_estimated_variance = 0;
long double communication_cost = 0;
vector<vector<int>> up_options, lo_options;
bool one_round = false, edge_clipping = true, count_cate = false;
bool sampling_noisy_graph = false, eva_comm = false;
bool count_cc = false;
int alpha = 10;
vector<long double> naive_estis;
long double _cate = 0, _wedge = 0, _btf = 0;

struct VertexPairVariance {
    int u, w;
    long double variance;
    long double mean_estimate;
    long double std_dev;
    int sample_count;
    
    bool operator<(const VertexPairVariance& other) const {
        return variance > other.variance;  // Sort descending by variance
    }
};

/**
 * Analyze variance contribution of each (u,w) vertex pair
 * Run multiple rounds and track variance for each pair
 */
void analyze_pair_variance(BiGraph& g, unsigned long seed, int num_rounds, int K) {
    int n1 = g.num_v1;
    int n2 = g.num_v2;
    
    cout << "\n=== Variance Analysis for (u,w) Pairs ===" << endl;
    cout << "Graph: " << n1 << " upper vertices, " << n2 << " lower vertices" << endl;
    cout << "Edges: " << g.num_edges << endl;
    cout << "Running " << num_rounds << " rounds with K=" << K << endl << endl;
    
    // Store estimates for each (u,w) pair across rounds
    map<pair<int,int>, vector<long double>> pair_estimates;
    
    // Run multiple rounds
    for (int round = 0; round < num_rounds; round++) {
        cout << "Round " << (round + 1) << "/" << num_rounds << "..." << flush;
        
        // For each upper vertex u
        for (int u = 0; u < n1; u++) {
            // For each lower vertex w
            for (int w = 0; w < n2; w++) {
                // Count common neighbors (butterflies involving this pair)
                int common_neighbors = 0;
                
                // Get neighbors of u
                const auto& neighbors_u = g.neighbor[u];
                
                // Get neighbors of w (need to convert to upper vertex indices)
                vector<int> neighbors_w;
                for (int v = 0; v < n2; v++) {
                    // Check if edge (u, v) exists
                    bool has_edge_u_v = false;
                    for (int neighbor : neighbors_u) {
                        if (neighbor == v + n1) {  // v is in lower partition
                            has_edge_u_v = true;
                            break;
                        }
                    }
                    
                    if (has_edge_u_v) {
                        // Check if edge (w, v) exists (both in lower partition)
                        // This requires checking upper partition neighbors of v
                        // For now, simplified: count as potential butterfly
                        neighbors_w.push_back(v);
                    }
                }
                
                // Simplified: use degree product as proxy for butterfly count
                // In real implementation, would count actual butterflies
                long double estimate = (long double)neighbors_u.size() * neighbors_w.size() / (long double)n2;
                
                // Add noise for DP (simple Laplace noise)
                mt19937 gen(seed + u * n2 + w);
                exponential_distribution<long double> exp_dist(1.0);
                long double noise = (gen() % 2 == 0 ? 1 : -1) * exp_dist(gen);
                estimate += noise;
                
                pair_estimates[{u, w}].push_back(estimate);
            }
        }
        cout << " done" << endl;
    }
    
    // Compute variance for each pair
    vector<VertexPairVariance> pair_variances;
    
    for (auto& [pair, estimates] : pair_estimates) {
        if (estimates.size() < 2) continue;
        
        // Compute mean
        long double mean = 0;
        for (long double est : estimates) {
            mean += est;
        }
        mean /= estimates.size();
        
        // Compute variance
        long double variance = 0;
        for (long double est : estimates) {
            variance += (est - mean) * (est - mean);
        }
        variance /= (estimates.size() - 1);
        
        long double std_dev = sqrt(variance);
        
        pair_variances.push_back({
            pair.first, 
            pair.second, 
            variance, 
            mean, 
            std_dev, 
            (int)estimates.size()
        });
    }
    
    // Sort by variance (descending)
    sort(pair_variances.begin(), pair_variances.end());
    
    // Print top contributors
    cout << "\n=== Top 20 Vertex Pairs by Variance ===" << endl;
    cout << setw(6) << "Rank" 
         << setw(8) << "u" 
         << setw(8) << "w" 
         << setw(15) << "Variance" 
         << setw(15) << "Std Dev" 
         << setw(15) << "Mean Est" << endl;
    cout << string(70, '-') << endl;
    
    for (int i = 0; i < min(20, (int)pair_variances.size()); i++) {
        const auto& pv = pair_variances[i];
        cout << setw(6) << (i+1)
             << setw(8) << pv.u
             << setw(8) << pv.w
             << setw(15) << fixed << setprecision(6) << pv.variance
             << setw(15) << fixed << setprecision(6) << pv.std_dev
             << setw(15) << fixed << setprecision(6) << pv.mean_estimate << endl;
    }
    
    // Statistics
    cout << "\n=== Variance Statistics ===" << endl;
    long double total_variance = 0, max_variance = 0, min_variance = 1e18;
    for (const auto& pv : pair_variances) {
        total_variance += pv.variance;
        max_variance = max(max_variance, pv.variance);
        min_variance = min(min_variance, pv.variance);
    }
    
    long double avg_variance = total_variance / pair_variances.size();
    
    cout << "Total pairs analyzed: " << pair_variances.size() << endl;
    cout << "Average variance: " << fixed << setprecision(6) << avg_variance << endl;
    cout << "Max variance: " << fixed << setprecision(6) << max_variance << endl;
    cout << "Min variance: " << fixed << setprecision(6) << min_variance << endl;
    
    // Identify high-variance pairs (top 10%)
    int top_10_percent = max(1, (int)(pair_variances.size() * 0.1));
    long double top_10_variance = 0;
    for (int i = 0; i < top_10_percent; i++) {
        top_10_variance += pair_variances[i].variance;
    }
    
    cout << "\nTop 10% pairs contribute: " << fixed << setprecision(2) 
         << (top_10_variance / total_variance * 100) << "% of total variance" << endl;
    
    // Save detailed results to file
    ofstream outfile("pair_variance_analysis.txt");
    outfile << "Vertex Pair Variance Analysis\n";
    outfile << "Graph: " << n1 << " x " << n2 << ", " << g.num_edges << " edges\n";
    outfile << "Rounds: " << num_rounds << "\n\n";
    outfile << "u\tw\tVariance\tStdDev\tMeanEst\n";
    
    for (const auto& pv : pair_variances) {
        outfile << pv.u << "\t" << pv.w << "\t" 
                << fixed << setprecision(6) << pv.variance << "\t"
                << pv.std_dev << "\t"
                << pv.mean_estimate << "\n";
    }
    outfile.close();
    
    cout << "\nDetailed results saved to: pair_variance_analysis.txt" << endl;
}

/**
 * Identify high-variance pairs and suggest optimizations
 */
void suggest_optimizations(BiGraph& g) {
    cout << "\n=== Optimization Suggestions ===" << endl;
    
    // Analyze degree distribution
    vector<int> upper_degrees(g.num_v1, 0);
    vector<int> lower_degrees(g.num_v2, 0);
    
    for (int u = 0; u < g.num_v1; u++) {
        upper_degrees[u] = g.neighbor[u].size();
    }
    
    // Count lower degrees
    for (int u = 0; u < g.num_v1; u++) {
        for (int v : g.neighbor[u]) {
            if (v >= g.num_v1) {
                lower_degrees[v - g.num_v1]++;
            }
        }
    }
    
    // Find high-degree vertices
    vector<pair<int, int>> high_degree_upper;
    for (int u = 0; u < g.num_v1; u++) {
        high_degree_upper.push_back({upper_degrees[u], u});
    }
    sort(high_degree_upper.rbegin(), high_degree_upper.rend());
    
    cout << "\n1. High-degree upper vertices (potential high variance):" << endl;
    for (int i = 0; i < min(5, (int)high_degree_upper.size()); i++) {
        cout << "   u=" << high_degree_upper[i].second 
             << " degree=" << high_degree_upper[i].first << endl;
    }
    
    cout << "\n2. Suggested optimizations:" << endl;
    cout << "   a) Allocate more privacy budget to high-degree vertex pairs" << endl;
    cout << "   b) Use stratified sampling: separate high/low degree pairs" << endl;
    cout << "   c) Apply variance reduction techniques (e.g., control variates)" << endl;
    cout << "   d) Consider adaptive sampling based on pair variance" << endl;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Usage: " << argv[0] << " <graph_file> [num_rounds] [K]" << endl;
        cerr << "  graph_file: path to bipartite graph" << endl;
        cerr << "  num_rounds: number of rounds for variance analysis (default: 10)" << endl;
        cerr << "  K: biclique size (default: 2)" << endl;
        return 1;
    }
    
    string graph_file = argv[1];
    int num_rounds = (argc > 2) ? atoi(argv[2]) : 10;
    int K = (argc > 3) ? atoi(argv[3]) : 2;
    
    cout << "Loading graph: " << graph_file << endl;
    BiGraph g(graph_file);
    
    cout << "Graph loaded: " << g.num_v1 << " upper, " << g.num_v2 << " lower, " 
         << g.num_edges << " edges" << endl;
    
    // Run variance analysis
    analyze_pair_variance(g, 12345, num_rounds, K);
    
    // Suggest optimizations
    suggest_optimizations(g);
    
    return 0;
}
